packages/range/src/repeat.ts

```diff
@@ -23,60 +23,68 @@ export function repeat<T>(
   mapFn: (i: number) => T,
   options: { fallback?: Accessor<T> } = {},
 ): Accessor<T[]> {
-  let disposers: (() => void)[] = [],
-    items: T[] = [],
-    prevLen = 0;
-
-  onCleanup(() => disposers.forEach(f => f()));
+  let prev: readonly T[] = [];
+  let prevLen = 0;
+  const disposers: (() => void)[] = [];
+  onCleanup(() => {
+    for (let index = 0; index < disposers.length; index++) {
+      disposers[index]!();
+    }
+  });
 
-  const mapLength = (len: number): T[] => {
-    if (len === 0) {
-      disposers.forEach(f => f());
+  // Truncate toward zero and force positive
+  const memoLen = createMemo(() => Math.max(times() | 0, 0));
 
-      if (options.fallback)
-        return createRoot(dispose => {
-          disposers = [dispose];
-          return (items = [options.fallback!()]);
-        });
+  return function mapLength(): T[] {
+    const len = memoLen();
+    if (len === prevLen) return prev as T[];
 
-      disposers = [];
-      return (items = []);
+    // Dispose of fallback or unnecessarry elements
+    if (prevLen === 0) disposers[0]?.();
+    else {
+      for (let index = len; index < disposers.length; index++) {
+        disposers[index]!();
+      }
     }
 
-    if (prevLen === 0) {
-      // after fallback case:
-      if (disposers[0]) disposers[0]();
-      for (let i = 0; i < len; i++) items[i] = createRoot(mapper.bind(void 0, i));
-      return items;
-    }
+    // The following prefers to use `prev.slice` to
+    // preserve any array element kind optimizations
+    // the runtime has made.
 
-    {
-      const diff = prevLen - len;
-      if (diff > 0) {
-        for (let i = prevLen - 1; i >= len; i--) disposers[i]!();
-        items.splice(len, diff);
-        disposers.splice(len, diff);
-        return items;
+    if (len === 0) {
+      const fallback = options.fallback;
+      if (fallback) {
+        // Show fallback if available
+        const next = prev.slice(0, 1);
+        next[0] = createRoot(dispose => {
+          disposers[0] = dispose;
+          return fallback();
+        });
+
+        disposers.length = 1;
+        prevLen = 0;
+        return (prev = next);
+      } else {
+        // Show empty array, otherwise
+        disposers.length = 0;
+        prevLen = 0;
+        return (prev = prev.slice(0, 0));
       }
     }
 
-    for (let i = prevLen; i < len; i++) items[i] = createRoot(mapper.bind(void 0, i));
-    return items;
-  };
+    const next = prev.slice(0, len);
 
-  const mapper = (index: number, dispose: () => void): T => {
-    disposers[index] = dispose;
-    return mapFn(index);
-  };
+    // Create new elements as needed
+    for (let index = prevLen; index < len; index++) {
+      next[index] = createRoot(dispose => {
+        disposers[index] = dispose;
+        return mapFn(index);
+      });
+    }
 
-  const memoLen = createMemo(() => Math.floor(Math.max(times(), 0)));
-  return () => {
-    const len = memoLen();
-    return untrack(() => {
```

Comment:
I removed this `untrack` call+closure, as far as I can tell it's redundant: `createRoot` already does both Owner and Listener.